test/functional/lua/command_line_completion_spec.lua

```diff
@@ -0,0 +1,166 @@
+    eq(
+      {
+        'vim["api"].nvim_buf_set_lines',
+        'vim["api"].nvim_buf_set_option'
+      },
+      get_completions('vim["api"].nvim_buf_', {
+        vim = {
+          api = {
+            nvim_buf_set_lines = true,
+            nvim_buf_set_option = true,
+            nvim_win_doesnt_match = true,
+          },
+          other_key = true,
+        }
+      })
+    )
+  end)
+
+  it('should return keys for string reffed dictionaries', function()
+    eq(
+      {
+        'vim["nested"]["api"].nvim_buf_set_lines',
+        'vim["nested"]["api"].nvim_buf_set_option'
+      },
+      get_completions('vim["nested"]["api"].nvim_buf_', {
+        vim = {
+          nested = {
+            api = {
+              nvim_buf_set_lines = true,
+              nvim_buf_set_option = true,
+              nvim_win_doesnt_match = true,
+            },
+          },
+          other_key = true,
+        }
+      })
+    )
+  end)
+
+  it('should be able to interpolate globals', function()
+    eq(
+      {
+        'vim[MY_VAR].nvim_buf_set_lines',
+        'vim[MY_VAR].nvim_buf_set_option'
+      },
+      get_completions('vim[MY_VAR].nvim_buf_', {
+        MY_VAR = "api",
+        vim = {
+          api = {
+            nvim_buf_set_lines = true,
+            nvim_buf_set_option = true,
+            nvim_win_doesnt_match = true,
+          },
+          other_key = true,
+        }
+      })
+    )
+  end)
+
+  it('should return everything if the input is of length 0', function()
+    eq({"other", "vim"}, get_completions('', { vim = true, other = true }))
+  end)
+
+  describe('get_parts', function()
+    it('should return an empty list for no separators', function()
+      eq({{}, 1}, get_compl_parts("vim"))
+    end)
+
+    it('just the first item before a period', function()
+      eq({{"vim"}, 5}, get_compl_parts("vim.ap"))
+    end)
+
+    it('should return multiple parts just for period', function()
+      eq({{"vim", "api"}, 9}, get_compl_parts("vim.api.nvim_buf"))
+    end)
+
+    it('should be OK with colons', function()
+      eq({{"vim", "api"}, 9}, get_compl_parts("vim:api.nvim_buf"))
+    end)
+
+    it('should work for just one string ref', function()
+      eq({{"vim", "api"}, 12}, get_compl_parts("vim['api'].nvim_buf"))
+    end)
+
+    it('should work for just one string ref, with double quote', function()
+      eq({{"vim", "api"}, 12}, get_compl_parts('vim["api"].nvim_buf'))
+    end)
+
+    it('should allows back-to-back string ref', function()
+      eq({{"vim", "nested", "api"}, 22}, get_compl_parts('vim["nested"]["api"].nvim_buf'))
+    end)
+
+    it('should allows back-to-back string ref with spaces before and after', function()
+      eq({{"vim", "nested", "api"}, 25}, get_compl_parts('vim[ "nested"  ]["api"].nvim_buf'))
+    end)
+
+    it('should allow VAR style loolup', function()
+      eq({{"vim", {"NESTED"}, "api"}, 20}, get_compl_parts('vim[NESTED]["api"].nvim_buf'))
+    end)
+  end)
+end)
```

Comment:
we also want an integration test for `ex_getln.c` integration. Could look into `viml/completion_spec.lua` and how it does screen tests.